ngraph/core/src/op/einsum.cpp

```diff
@@ -0,0 +1,277 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <algorithm>
+#include <cctype>
+#include <ngraph/validation_util.hpp>
+#include <regex>
+#include <string>
+#include <unordered_map>
+
+#include "itt.hpp"
+#include "ngraph/op/einsum.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+NGRAPH_RTTI_DEFINITION(op::v7::Einsum, "Einsum", 7);
+
+op::v7::Einsum::Einsum(const OutputVector& inputs, const std::string& equation)
+    : Op(inputs)
+    , m_equation(equation)
+{
+    constructor_validate_and_infer_types();
+}
+
+bool op::v7::Einsum::parse_equation(std::vector<std::string>& input_subscripts,
+                                    std::string& output_subscript)
+{
+    // remove extra white-spaces from the equation
+    m_equation.erase(std::remove_if(m_equation.begin(), m_equation.end(), ::isspace),
+                     m_equation.end());
+
+    // split equation to input subscripts and an output subscript
+    auto pos_output_delimeter = m_equation.find("->");
+    auto input_subscripts_str = m_equation.substr(0, pos_output_delimeter);
+
+    // split the input subscripts into a vector of input subscripts
+    input_subscripts.clear();
+    std::istringstream input;
+    input.str(input_subscripts_str);
+    for (std::string input_subscript; std::getline(input, input_subscript, ',');)
+    {
+        input_subscripts.push_back(input_subscript);
+    }
+
+    // check that each input subscript contains only alphabetic letter or ellipsis
+    const std::regex subscript_regex("[A-Za-z]*|[A-Za-z]*\\.\\.\\.[A-Za-z]*");
+    for (auto const& input_subscript : input_subscripts)
+    {
+        if (std::regex_match(input_subscript, subscript_regex) == false)
+        {
+            return false;
+        }
+    }
+
+    // figure out if ellipsis is met in input subscripts
+    const std::regex ellipsis_regex("\\.\\.\\.");
+    bool is_ellipsis_met = false;
+    std::smatch ellipsis_match;
+    for (auto const& input_subscript : input_subscripts)
+    {
+        if (std::regex_search(input_subscript, ellipsis_match, ellipsis_regex) == true)
```

Comment:
may be if (std::regex_search(input_subscript, ellipsis_match, ellipsis_regex))

Reply:
No, I intentionally do it in order to make sure that I use a function that returns boolean type. And there is nothing bad in this explicit form.